File: lib/data/models/pomodoro_task_model.dart
import 'package:get/get.dart';

import 'package:sptimer/data/enums/pomodoro_status.dart';
import 'package:sptimer/data/enums/timer_status.dart';
import 'package:sptimer/data/enums/tones.dart';

const _kRemainingDuration = 'kCurrentRemainingDurationKey';
const _kPomodoroRound = 'PomodoroRoundKey';
const _kPomodoroStatus = 'kPomodoroStatus';
const _kMaxRound = 'kMaxRoundKey';
const _kWorkDuration = 'kWorkDurationKey';
const _kShortBreakDuration = 'kShortBreakDurationKey';
const _kLongBreakDuration = 'kLongBreakDurationKey';
const _kTimerStatus = 'kTimerStatus';
const _kTitle = 'kTitle';
const _kTaskId = 'kTaskId';
const _kVibrate = 'kVibrate';
const _kTone = 'kTone';
const _kToneVolume = 'kToneVolume';
const _kStatusVolume = 'kStatusVolume';
const _kReadStatusAloud = 'kReadStatusAloud';

class PomodoroTaskModel {
  PomodoroTaskModel({
    this.remainingDuration,
    this.id,
    this.pomodoroRound = 0,
    this.pomodoroStatus = PomodoroStatus.work,
    this.timerStatus = TimerStatus.cancel,
    required this.title,
    required this.workDuration,
    required this.shortBreakDuration,
    required this.longBreakDuration,
    required this.maxPomodoroRound,
    required this.vibrate,
    required this.tone,
    required this.toneVolume,
    required this.statusVolume,
    required this.readStatusAloud,
  });

  final Duration? remainingDuration;
  final int? id;
  final String title;
  final Duration workDuration;
  final Duration shortBreakDuration;
  final Duration longBreakDuration;
  final int maxPomodoroRound;
  final int pomodoroRound;
  final PomodoroStatus pomodoroStatus;
  final TimerStatus timerStatus;
  final bool vibrate;
  final Tones tone;
  final double toneVolume;
  final double statusVolume;
  final bool readStatusAloud;

  Map<String, dynamic> toMap() => {
        _kRemainingDuration: remainingDuration?.inSeconds,
        _kWorkDuration: workDuration.inSeconds,
        _kShortBreakDuration: shortBreakDuration.inSeconds,
        _kLongBreakDuration: longBreakDuration.inSeconds,
        _kPomodoroRound: pomodoroRound,
        _kMaxRound: maxPomodoroRound,
        _kPomodoroStatus: pomodoroStatus.index,
        _kTimerStatus: timerStatus.index,
        _kTitle: title,
        _kTaskId: id,
        _kVibrate: vibrate,
        _kTone: '${tone.name}.${tone.type}',
        _kToneVolume: toneVolume,
        _kStatusVolume: statusVolume,
        _kReadStatusAloud: readStatusAloud,
      };

  static PomodoroTaskModel fromMap(Map<dynamic, dynamic> data) {
    final tone = Tones.values.firstWhere((element) {
      final toneString = (data[_kTone] as String);
      final formatIndex = toneString.indexOf('.');
      final name = toneString.substring(0, formatIndex);
      return element.name == name;
    });
    return PomodoroTaskModel(
      remainingDuration: (data[_kRemainingDuration] as int?)?.seconds,
      workDuration: (data[_kWorkDuration] as int).seconds,
      shortBreakDuration: (data[_kShortBreakDuration] as int).seconds,
      longBreakDuration: (data[_kLongBreakDuration] as int).seconds,
      pomodoroRound: data[_kPomodoroRound],
      maxPomodoroRound: data[_kMaxRound]!,
      pomodoroStatus: PomodoroStatus.values[data[_kPomodoroStatus]],
      timerStatus: TimerStatus.values[data[_kTimerStatus]],
      title: data[_kTitle],
      id: data[_kTaskId],
      vibrate: data[_kVibrate],
      tone: tone,
      toneVolume: data[_kToneVolume],
      statusVolume: data[_kStatusVolume],
      readStatusAloud: data[_kReadStatusAloud],
    );
  }

  PomodoroTaskModel copyWith({
    Duration? remainingDuration,
    Duration? currentMaxDuration,
    int? id,
    String? title,
    Duration? workDuration,
    Duration? shortBreakDuration,
    Duration? longBreakDuration,
    int? maxPomodoroRound,
    int? pomodoroRound,
    PomodoroStatus? pomodoroStatus,
    TimerStatus? timerStatus,
    bool? vibrate,
    Tones? tone,
    double? toneVolume,
    double? statusVolume,
    bool? readStatusAloud,
  }) {
    return PomodoroTaskModel(
      remainingDuration: remainingDuration ?? this.remainingDuration,
      id: id ?? this.id,
      title: title ?? this.title,
      workDuration: workDuration ?? this.workDuration,
      shortBreakDuration: shortBreakDuration ?? this.shortBreakDuration,
      longBreakDuration: longBreakDuration ?? this.longBreakDuration,
      maxPomodoroRound: maxPomodoroRound ?? this.maxPomodoroRound,
      pomodoroRound: pomodoroRound ?? this.pomodoroRound,
      pomodoroStatus: pomodoroStatus ?? this.pomodoroStatus,
      timerStatus: timerStatus ?? this.timerStatus,
      vibrate: vibrate ?? this.vibrate,
      tone: tone ?? this.tone,
      toneVolume: toneVolume ?? this.toneVolume,
      statusVolume: statusVolume ?? this.statusVolume,
      readStatusAloud: readStatusAloud ?? this.readStatusAloud,
    );
  }
}

File: android/app/src/main/java/com/example/pomotimer/models/PomodoroTaskModel.java
package com.example.pomotimer.models;

import com.example.pomotimer.Constants;
import com.example.pomotimer.timer.PomodoroStatus;
import com.example.pomotimer.timer.TimerStatus;

import java.util.HashMap;
import java.util.Map;

public class PomodoroTaskModel {
    public PomodoroTaskModel(
            int remainingDuration,
            int id,
            String title,
            int workDuration,
            int shortBreakDuration,
            int longBreakDuration,
            int maxPomodoroRound,
            int pomodoroRound,
            PomodoroStatus pomodoroStatus,
            TimerStatus timerStatus,
            boolean vibrate,
            String toneName,
            double toneVolume,
            double statusVolume,
            boolean readStatusAloud
    ) {
        this.remainingDuration = remainingDuration;
        this.id = id;
        this.title = title;
        this.workDuration = workDuration;
        this.shortBreakDuration = shortBreakDuration;
        this.longBreakDuration = longBreakDuration;
        this.maxPomodoroRound = maxPomodoroRound;
        this.pomodoroRound = pomodoroRound;
        this.pomodoroStatus = pomodoroStatus;
        this.timerStatus = timerStatus;
        this.vibrate = vibrate;
        this.toneName = toneName;
        this.toneVolume = toneVolume;
        this.statusVolume = statusVolume;
        this.readStatusAloud = readStatusAloud;
    }

    public int remainingDuration;
    public int id;
    public String title;
    public int workDuration;
    public int shortBreakDuration;
    public int longBreakDuration;
    public int maxPomodoroRound;
    public int pomodoroRound;
    public PomodoroStatus pomodoroStatus;
    public TimerStatus timerStatus;
    public boolean vibrate;
    public String toneName;
    public double toneVolume;
    public double statusVolume;
    public boolean readStatusAloud;


    public Map<String, Object> toMap() {
        Map<String, Object> map = new HashMap<>();
        map.put(Constants.kRemainingDuration, remainingDuration);
        map.put(Constants.kWorkDuration, workDuration);
        map.put(Constants.kShortBreakDuration, shortBreakDuration);
        map.put(Constants.kLongBreakDuration, longBreakDuration);
        map.put(Constants.kPomodoroRound, pomodoroRound);
        map.put(Constants.kMaxRound, maxPomodoroRound);
        map.put(Constants.kPomodoroStatus, pomodoroStatus.ordinal());
        map.put(Constants.kTimerStatus, timerStatus.ordinal());
        map.put(Constants.kTitle, title);
        map.put(Constants.kTaskId, id);
        map.put(Constants.kVibrate, vibrate);
        map.put(Constants.kTone, toneName);
        map.put(Constants.kToneVolume, toneVolume);
        map.put(Constants.kStatusVolume, statusVolume);
        map.put(Constants.kReadStatusAloud, readStatusAloud);
        return map;
    }


    public static PomodoroTaskModel fromMap(Map<String, Object> data) {
        return new PomodoroTaskModel(
                (int) data.get(Constants.kRemainingDuration),
                (int) data.get(Constants.kTaskId),
                (String) data.get(Constants.kTitle),
                (int) data.get(Constants.kWorkDuration),
                (int) data.get(Constants.kShortBreakDuration),
                (int) data.get(Constants.kLongBreakDuration),
                (int) data.get(Constants.kMaxRound),
                (int) data.get(Constants.kPomodoroRound),
                PomodoroStatus.values()[(int) data.get(Constants.kPomodoroStatus)],
                TimerStatus.values()[(int) data.get(Constants.kTimerStatus)],
                (Boolean) data.get(Constants.kVibrate),
                (String) data.get(Constants.kTone),
                (Double) data.get(Constants.kToneVolume),
                (Double) data.get(Constants.kStatusVolume),
                (Boolean) data.get(Constants.kReadStatusAloud)
        );
    }


}
